.build/build

```
@@ -0,0 +1,8 @@
+#!/bin/sh -ef
+
+cd $(dirname "$0")/..
```

See https://www.shellcheck.net/

<pre><code>Line 3:
cd $(dirname "$0")/..
   ^-- <a href="https://www.shellcheck.net/wiki/SC2046">SC2046</a> (warning): Quote this to prevent word splitting.</code></pre>

```suggestion
cd "$(dirname "$0")"/..
```

---

Actually, shellcheck seems to be wrong here. There are 2 shells:

* Internal one executes `dirname "$0"`, and this one indeed requires quoting, otherwise `$0` gets expanded first and might result in more than 1 argument being sent over to dirname.
* External one executes `cd ` + result from internal shell + `/..`. Parsing and tokenization of external one actually gets decided first (otherwise, you won't be able to even derive that there is internal shell call), so it's already been decided that there's just 2 tokens: 1st is `cd` and 2nd will be decided later as result of execution of that expression.

So, there's no real need to quote it.

Quick test: I'm in the following path: `/home/greycat/z/x  y/a b` (note two spaces between `x  y` and one space between `a b`).

```
$ pwd | hd   
00000000  2f 68 6f 6d 65 2f 67 72  65 79 63 61 74 2f 7a 2f  |/home/greycat/z/|
00000010  78 20 20 79 2f 61 20 62  0a                       |x  y/a b.|
$ cat foo.sh
#!/bin/sh -e

pwd
cd "$(dirname "$0")"/..
pwd
$ ./foo.sh
/home/greycat/z/x  y/a b
/home/greycat/z/x  y
```

For sanity checking, just checked it with dash (=sh), bash, zsh.

---

> Actually, shellcheck seems to be wrong here.

I don't think so.

> So, there's no real need to quote it.
> 
> Quick test: I'm in the following path: `/home/greycat/z/x y/a b` (note two spaces between `x y` and one space between `a b`).

This test doesn't show anything for 2 reasons - according to the `$ cat foo.sh` output that you did there, you were already testing the corrected solution with quoted `"$(dirname "$0")"`, so no wonder you can't find any issues with it.

The other reason becomes obvious when you actually print the contents of the `$0` variable and the result of `dirname "$0"`, for example like this (and also notice that now I've actually removed the quotes in `cd $(dirname "$0")`, so we'll be able to see how it's wrong):

```console
pp@DESKTOP-89OPGF3:~/x y/a b$ cat foo.sh
#!/bin/sh -e

printf %s "$0" | hd
printf %s "$(dirname "$0")" | hd

pwd
cd $(dirname "$0")/..
pwd
```

If you run the script as `./foo.sh`:

```console
pp@DESKTOP-89OPGF3:~/x y/a b$ ./foo.sh
00000000  2e 2f 66 6f 6f 2e 73 68                           |./foo.sh|
00000008
00000000  2e                                                |.|
00000001
/home/pp/x y/a b
/home/pp/x y
```

You can see that the `$0` contained exactly the path `./foo.sh` that the script was run with, and `dirname "$0"` is `.` (just 1-character string, the dot) - since `.` doesn't contain any spaces, the unquoted version indeed works in this case.

However, if the invocation path includes a directory name with a space, the unquoted version stops working (as expected; this is a shell script where you really have to quote almost everything, I didn't make it up):

```console
pp@DESKTOP-89OPGF3:~/x y$ ./a\ b/foo.sh
00000000  2e 2f 61 20 62 2f 66 6f  6f 2e 73 68              |./a b/foo.sh|
0000000c
00000000  2e 2f 61 20 62                                    |./a b|
00000005
/home/pp/x y
./a b/foo.sh: 7: cd: can't cd to ./a
```

---

Uhh, you're absolutely correct, sorry for all the confusion. My bad for not noticing it. Let me apply the fix.